src/estimation.jl

```diff
@@ -38,4 +40,157 @@ function esprit(x::AbstractArray, M::Integer, p::Integer, Fs::Real=1.0)
 end
 
+"""
+    jacobsen(x::Vector, Fs::Real = 1.0)
+
+Estimate the largest frequency (in Hz) in the signal `x` using Jacobsen's
+algorithm [^Jacobsen2007].
+
+[^Jacobsen2007]: E Jacobsen and P Kootsookos, "Fast, Accurate Frequency Estimators", Chapter
+10 in "Streamlining Digital Signal Processing", edited by R. Lyons, 2007, IEEE Press.
+"""
+function jacobsen(x::Vector{<:Real}, Fs::Real = 1.0)
+    N = length(x)
+    X = rfft(x)
+    k = argmax(abs.(X))  # index of DFT peak
+    if ((k != div(N,2)+1) && (k != 1))  # avoid out-of-bounds indexing
+        # Jacoben's formula
+        δ = -real((X[k+1] - X[k-1]) / (2X[k] - X[k-1] - X[k+1]) )
+    else
+        δ = 0.0
+    end
+
+    return (k - 1 + δ)*Fs/N
+end
+
+function jacobsen(x::Vector{<:Complex}, Fs::Real = 1.0)
+    N = length(x)
+    X = fftshift(fft(x))
+    k = argmax(abs.(X))  # index of DFT peak
+    fpeak = fftshift(fftfreq(N, Fs))[k]  # peak frequency
+    if ((k != N) && (k != 1))  # avoid out-of-bounds indexing
+        # jacobsen's formula
+        δ = -real((X[k+1] - X[k-1]) / (2X[k] - X[k-1] - X[k+1]) )
+    else
+        δ = 0.0
+    end
+
+    return fpeak + δ*Fs/N
+end
+
+"""
+    quinn(x::Vector, f0::Real, Fs::Real = 1.0 ; tol = 1e-6, maxiters = 20)
+
+    quinn(x::Vector, Fs::Real = 1.0 ; kwargs...)
+
+    quinn(x::Vector ; kwargs...)
+
+Algorithms by Quinn and Quinn & Fernandes for frequency estimation. Given a
+signal `x` and an initial guess `f0`, estimate and return the frequency of the
+largest sinusoid in `x`. `Fs` is the sampling frequency. All frequencies are
+expressed in Hz.
+
+If the initial guess `f0` is not provided, then a guess is calculated using
+Jacobsen's estimator. If the sampling frequency `Fs` is not provided, then it
+is assumed that `Fs = 1.0`.
+
+The following keyword arguments control the algorithm's behavior:
+
+- `tol`: () the algorithm stops when the absolut value of the
+  difference between two consecutive estimates is less than `tol`. Defaults to
+  `1e-6`.
+- `maxiters`: the maximum number of iterations to run. Defaults to `20`.
+
+Returns a tuple `(estimate, reachedmaxiters)`, where `estimate` is the
+estimated frequency, and `reachedmaxiters` is `true` if the algorithm finished
+after running for `maxiters` iterations (this may indicate that the algorithm
+did not converge).
+
+If the signal `x` is real, then the algorithm used is [^Quinn1991]. If the signal is
+complex, the algorithm is [^Quinn2009].
+
+[^Quinn1991]: B Quinn and J Fernandes, "A fast efficient technique for the
+estimation of frequency", Biometrika, Vol. 78 (1991).
+
+[^Quinn2009]: B Quinn, "Recent advances in rapid frequency estimation", Digital
+Signal Processing, Vol. 19 (2009), Elsevier.
+
+"""
+quinn(x ; kwargs...) = quinn(x, jacobsen(x, 1.0), 1.0 ; kwargs...)
+
+quinn(x, Fs ; kwargs...) = quinn(x, jacobsen(x, Fs), Fs ; kwargs...)
+
+function quinn(x::Vector{<:Real}, f0::Real, Fs::Real ; tol = 1e-6, maxiters = 20)
+    fₙ = Fs/2
+    T = length(x)
```

Comment:
Nitpick, but for `jacobsen`, this is called `N`, which I'd prefer here, too, not only for consistency, but also because `T` suggests either a time/duration in seconds or a -- in the julia context -- a type.

Reply:
The reason `T` is used here is to follow the paper's nomenclature. The code mimics all variables in the paper (for example `ξ`, `β`, etcetera). I think this makes it easier to study the code and the paper side-by-side.

Having said that, I agree that in this case `T` may be too confusing. I'll make this change.